src/text.rs

```diff
@@ -434,65 +427,108 @@ impl TextBox {
             self.key(bounds, zoom),
         );
 
-        if let Some(start_cursor) = buffer.hit(
-            select_start.0 - screen_position.0,
-            select_start.1 - screen_position.1,
-        ) {
-            if let Some(end_cursor) = buffer.hit(
-                select_end.0 - screen_position.0,
-                select_end.1 - screen_position.1,
-            ) {
-                if start_cursor.index == end_cursor.index {
-                    return (vec![], String::new());
+        let (start_cursor, end_cursor, start_y, end_y) = match &selection.selection {
+            SelectionKind::Drag { mut start, mut end } => {
+                if start.1 > end.1 || start.0 > end.0 {
+                    std::mem::swap(&mut start, &mut end);
+                }
+                if screen_position.1 > end.1 || screen_position.1 + bounds.1 < start.1 {
+                    return None;
                 }
 
-                let mut y = screen_position.1;
-                for line in buffer.layout_runs() {
-                    let line_contains =
-                        move |y_point: f32| y_point >= y && y_point <= y + line_height;
-                    if line_contains(select_start.1)
-                        || line_contains(select_end.1)
-                        || (select_start.1 < y && select_end.1 > y + line_height)
-                    {
-                        if let Some((highlight_x, highlight_w)) =
-                            line.highlight(start_cursor, end_cursor)
+                let start_cursor =
+                    buffer.hit(start.0 - screen_position.0, start.1 - screen_position.1)?;
+                let end_cursor =
+                    buffer.hit(end.0 - screen_position.0, end.1 - screen_position.1)?;
+                (start_cursor, end_cursor, start.1, end.1)
+            }
+            SelectionKind::Click { mode, position, .. } => {
+                let cursor = buffer.hit(
+                    position.0 - screen_position.0,
+                    position.1 - screen_position.1,
+                )?;
+
+                let line = buffer.lines.get(cursor.line)?;
+
+                match mode {
+                    SelectionMode::Word => {
+                        let text = line.text();
+                        if text
+                            .get(cursor.index..cursor.index)?
```

Comment:
I don't think this does anything since the `.get(cursor.index..cursor.index)` will always be an empty slice. I _think_ this check is to make sure that you can't do a word double-click selection when you're hovering over whitespace, but it still seems to try with the current impl where it winds up selecting words on either side minus the first character on the left/right side's word depending on the cursor's affinity (aka if it's on the left or right side of the whitespace character)

![image](https://github.com/Inlyne-Project/inlyne/assets/30302768/4f449afa-d9c1-4bf7-a3b1-109fa2e474ac)